source/utilities.tex

```diff
@@ -18341,10 +18341,27 @@
 if there is a template type argument $\tcode{B}_{i}$
 for which \tcode{bool($\tcode{B}_{i}$::value)} is \tcode{false},
 then instantiating \tcode{conjunction<$\tcode{B}_{1}$, $\dotsc$, $\tcode{B}_{N}$>::value}
-does not require the instantiation of \tcode{$\tcode{B}_{j}$::value} for $j > i$.
+does not require a definition of \tcode{$\tcode{B}_{j}$::value} to exist\iref{temp.inst} for $i < j \leq N$.
```

Comment:
As a reader of this, I'm left wondering if a declaration is required to exist.

Reply:
... and I think fixing that takes us out of editorial country. :(

Reply:
> (For LWG: is an implementation allowed to require later B_js to be complete types?)

This sentence is partially redundant with the blanket wording in http://eel.is/c++draft/meta#rqmts-5 that allows the templates in [meta] to be instantiated with arguments that are incomplete types. It might be clearer to specify this as "If there is a template type argument `B_i` for which `B_i::value` is `false`, then `B_j` must be a complete type for `0 <= j <= i`. [_Note:_ `B_j` can be an incomplete type for `i < j <= N`.-_end note_]"

Reply:
> If there is a template type argument `B_i` for which `B_i::value` is `false`, then `B_j` must be a complete type for `0 <= j <= i`.

That's wrong. We want `conjunction_v<falsy, incomplete-or-hard-error, falsy>` to be valid.

Reply:
Ah yes, we need to only talk about the first `Bi` for which `Bi::value` is `false`.

Reply:
Something like:

> For a specialization `conjunction<B_1 , ..., B_N>`, let `B_i` be the first type in `B` for which `B_i::value` is `false`, or `N` if there is no such type in `B`.  `B_j` shall be a complete type for `0 <= j <= i`. [*Note: ...

Reply:
Can we just say "the first type in `B`" like that, or does it need to be "the first type in `B_1` ... `B_N`"?

Reply:
How about this?

![conjunction](https://user-images.githubusercontent.com/1254480/58462105-1bb50080-8129-11e9-9e6a-6cbd87eb73e3.png)

Reply:
Or with the paragraph defining the base class moved earlier, and the bogus "shalls" in the last paragraph replaced (because that's a requirement on the implementation of `conjunction`):

![conjunction](https://user-images.githubusercontent.com/1254480/58463799-6e43ec00-812c-11e9-802e-30a10d2023a4.png)

I'm not very happy with the normative wording in paragraph 5 though.

Reply:
Yes, the "so" is too much explanation as opposed to specification.

Reply:
Also it should be `1 <= j <= i` in both cases, not `0 <= j <= i`.

How about just replacing "so" with a period, making it two normative sentences?

And maybe reordering it as:

Bj shall be a complete type for 1 <= j <= i. Instantiation of Bj::value is required for 1 <= j <= i. Instantiation of Bj::value is not required for i < j <= N.

Reply:
> Also it should be `1 <= j <= i` in both cases, not `0 <= j <= i`.

I think the `0` managed to sneak in because I was thinking about defining `B_0` to be `true_type`. We've notably lost the fact that `conjunction<>` derives from `true_type` in the recent formulations. I suggest we replace para 3 with:

> For a specialization `conjunction<B_1, ... , B_N>`, let `B_0` be `true_type` and `B_i` be the first type in `B_0, ... , B_N` for which `bool(B_i::value)` is `false` or `B_N` if there is no such type.

It then notably doesn't matter of we use `0` or `1` as the lower bound for the requirements in the following paragraphs since `true_type` certainly meets all such requirements.

> How about just replacing "so" with a period, making it two normative sentences?

"`foo` shall be a complete type" is a necessary condition when "Instantiation of `foo::value` is required", we could drop the "shall be a complete type" bit, or move it into a note if we want to be painfully clear. (I'm personally inclined to drop it: I don't think this is a case where adding more words improves clarity.)

In para 6, "shall have a member `value` which is convertible to `bool`" falls a bit short, e.g., `struct S { double& value; }` has a member `value` which is convertible to `bool`. I think we want to require `bool(B_i::value)` to be a constant expression.